hostdiscovery/hostdiscovery.go

```diff
@@ -0,0 +1,160 @@
+	_ "github.com/docker/swarm/discovery/token"
+)
+
+const defaultHeartbeat = 10
+
+// JoinCallback provides a callback event for new node joining the cluster
+type JoinCallback func(entries []net.IP)
+
+// LeaveCallback provides a callback event for node leaving the cluster
+type LeaveCallback func(entries []net.IP)
+
+// HostDiscovery primary interface
+type HostDiscovery interface {
+	StartDiscovery(*config.ClusterCfg, JoinCallback, LeaveCallback) error
+	StopDiscovery() error
+	Fetch() ([]net.IP, error)
+}
+
+type hostDiscovery struct {
+	discovery discovery.Discovery
+	nodes     []net.IP
+	stopChan  chan struct{}
+	sync.Mutex
+}
+
+// NewHostDiscovery function creates a host discovery object
+func NewHostDiscovery() HostDiscovery {
+	return &hostDiscovery{nodes: []net.IP{}, stopChan: make(chan struct{})}
+}
+
+func (h *hostDiscovery) StartDiscovery(cfg *config.ClusterCfg, joinCallback JoinCallback, leaveCallback LeaveCallback) error {
+	hb := cfg.Heartbeat
+	if hb == 0 {
+		hb = defaultHeartbeat
+	}
+	d, err := discovery.New(cfg.Discovery, hb)
+	if err != nil {
+		return err
+	}
+
+	go d.Watch(func(entries []*discovery.Entry) {
+		h.processCallback(entries, joinCallback, leaveCallback)
+	})
+
+	if ip := net.ParseIP(cfg.Address); ip == nil {
+		return errors.New("Address config should be either ipv4 or ipv6 address")
+	}
+
+	if err := d.Register(cfg.Address + ":0"); err != nil {
+		return err
+	}
+
+	h.Lock()
+	h.discovery = d
+	h.Unlock()
+
+	go sustainHeartbeat(d, hb, cfg, h.stopChan)
+	return nil
+}
+
+func (h *hostDiscovery) StopDiscovery() error {
+	h.Lock()
+	stopChan := h.stopChan
+	h.Unlock()
+
+	close(stopChan)
+	return nil
+}
+
+func sustainHeartbeat(d discovery.Discovery, hb uint64, config *config.ClusterCfg, stopChan chan struct{}) {
+	for {
+		time.Sleep(time.Duration(hb) * time.Second)
+
+		select {
+		case <-stopChan:
+			return
+		default:
+		}
+
+		if err := d.Register(config.Address + ":0"); err != nil {
+			log.Error(err)
+		}
+	}
+}
+
+func (h *hostDiscovery) processCallback(entries []*discovery.Entry, joinCallback JoinCallback, leaveCallback LeaveCallback) {
+	updated := hosts(entries)
+	h.Lock()
+	existing := h.nodes
+	h.Unlock()
+	added, removed := diff(existing, updated)
+	if len(added) > 0 {
+		joinCallback(added)
+	}
+	if len(removed) > 0 {
+		leaveCallback(removed)
+	}
+}
+
+func diff(existing []net.IP, updated []net.IP) (added []net.IP, removed []net.IP) {
+	updateSet := mapset.NewSet()
+	for _, ip := range updated {
+		updateSet.Add(ip.String())
+	}
+	existingSet := mapset.NewSet()
+	for _, ip := range existing {
+		existingSet.Add(ip.String())
+	}
+	addSlice := updateSet.Difference(existingSet).ToSlice()
+	removeSlice := existingSet.Difference(updateSet).ToSlice()
+	for _, ip := range addSlice {
+		added = append(added, net.ParseIP(ip.(string)))
+	}
+	for _, ip := range removeSlice {
+		removed = append(removed, net.ParseIP(ip.(string)))
+	}
+	return
```

Comment:
As a general comment why don't we always keep the host entries as a `Set` so that we don't need to create these temporary sets every time an update to the set happens. Then when an update you just need to keep that set and discard the old set. This is assuming that you always get the whole set in every update. If not, then this whole logic doesn't make sense any ways.